types/module/module.go

```diff
@@ -278,11 +278,11 @@ func (m *Manager) RegisterInvariants(ir sdk.InvariantRegistry) {
 // RegisterRoutes registers all module routes and module querier routes
 func (m *Manager) RegisterRoutes(router sdk.Router, queryRouter sdk.QueryRouter, legacyQuerierCdc *codec.LegacyAmino) {
 	for _, module := range m.Modules {
-		if !module.Route().Empty() {
-			router.AddRoute(module.Route())
+		if r := module.Route(); !r.Empty() {
```

Comment:
I don't see much benefit here. Using methods allow the cache to kicking in and you spare memory allocation. Not a big deal, it just looks a bit of a pointless change?

Reply:
Yeah I don't think these changes are necessary. Makes it generally harder to read. My rule of thumb is typically, if a value is used more than once, put it in a variable. I'd revert these changes.

Reply:
It is used more then once. See the line below. That's why I did that change. PRO TIP: see also the PR description.

Reply:
> PRO TIP: see also the PR description.

I saw it. Problem is that you didn't see my `Using methods allow the cache to kick in and you spare memory allocation.`.
My advice is read it again. And again, should that help.

Please revert.

Reply:
When imperative, not jitted methods are cached? Do you have any resource about it, especially related to Go? I would imagine that JIT could do it. But Go doesn't have a JIT.

Reply:
To me it looks like there is extra allocation as @robert-zaremba is pointing out. `Route()` allocates a few things under the hood. Maybe it's not such a big deal because this is just on app startup, but still I would do it the way @robert-zaremba based on all my past experience (mostly in other languages).

@alessio are you able to say a little bit more about how the cache would help? I'm not understanding this part.

Reply:
I was not referring to any Golang caching mechanism. I apologise for the lack of clarity as I understand it may have contributed to creating more confusion here.

When a program is executed though many things happen in the operating system and various levels of caching are involved, some of which live within the operating system itself, whilst others live within the hardware (e.g. CPU). What I meant is that it's very common for repeated calls that return exactly the same value to be cached elsewhere outside of the compiler's control. Hence allocating a new variable in this case does not give any extra guarantee of better performance.

All in all: IMvHO this change does not bring any demonstrable significant performance gain. It's just a cosmetic change and again IMvHO (and AFAIK) we have bigger fish to fry than fixing `golangci-lint` pedantic warnings - e.g. writing test cases (codecov.io still shows a depressing 54%), docs, so on and so forth.

My 2c.

Reply:
You're absolutely right, my bad. I think this change set is fine OK.

Reply:
Thanks @alexanderbez

Reply:
@alessio  - of course this is changeset is minor. Like I wrote, I did it because I was in this function while tracing an app and noticed it. If I wouldn't do the change with `Trim`, I would probably skip it. But it was few seconds to do that change. 
Also, I don't believe that any function call cache validator will be smarter and faster than just using this reference - it's local and it will stay in the CPU cache anyway. Statically compiled, imperative languages don't do that. Especially here the module has a complex object and it can't mutate. To use a runtime function call cache you will need to assert that nothing changed. JIT or pure languages could do that.

I don't understand why do you like to bash the work I'm doing. I was proving that error checking is kind of critical. It's easier to do few not necessary error checks then be cautious to not forget a crucial check or wrongly assigned error. That's why we have tools for that. During the SDK call when I mentioned about `errcheck` nobody was against. Instead of spending hours of yours and mine time criticizing you could also write test cases.

Reply:
**Disclaimer #1**: I've approved this PR already. Since the author seems to enjoy some pointless debate though, I feel compelled to entertain this circus for a tad longer.

**Disclaimer #2**: This is my last response in this thread. I have quite a number of things on my plate and I cannot afford any further meaningless distractions.

>  of course this is changeset is minor.

Is it? Sounds like a astdounding revelation. /me truly gobsmacked. 

> I don't believe that any function call cache validator will be smarter and faster than just using this reference

Whether you believe it or not does not change that you can't predict what happens at runtime beforehand, once it's compiled on either your machine, my machine or someone else's machine. __Fact__.

> it's local and it will stay in the CPU cache anyway

That too proves another point of mine (just-a-cosmetic-change). Bingo (once more).

> Statically compiled, imperative languages don't do that - even on hardware, the the processor will need to know the whole context of the call. Especially here the module has a complex object and it can't mutate.

Again, this proves nothing for you can't predict on all target platforms what happens at runtime unless you dig into the binary's assembly **for each platform this tiny, insignificant piece of code will be compiled**.

I approved this PR. I warmly advise against tilting at windmills. There is a whole ambitious [Cosmos SDK testing tools & sandbox](https://paper.dropbox.com/doc/Cosmos-SDK-testing-tools-sandbox--A~aBHVGpDiL3SL6jC7zbV~moAg-x9w7WdAbfipHeNvFaOu0l) work plan drafted by this PR's author that will hardly be implemented by just staring at it.

'Been fun.

    AT

Reply:
> That too proves another point of mine (just-a-cosmetic-change). Bingo (once more).

It is not. Don't take my fragments without a context. `r := funWithComplexDependencies(); doSomeithg(r)` is totally different in not-jitted, imperative languages than `funWithComplexDependencies(); doSomething(funWithComplexDependencies())`.